src/mono/mono/component/mini-wasi-debugger.c

```diff
@@ -0,0 +1,149 @@
+#ifdef HOST_WASI
+#include <glib.h>
+#include <mono/mini/debugger-agent-external.h>
+#include <mono/component/debugger-engine.h>
+#include <mono/metadata/components.h>
+
+static int conn_fd;
+static int log_level = 1;
+static int retry_receive_message = 5000;
+
+__attribute__((import_module("wasi_snapshot_preview1")))
+__attribute__((import_name("sock_accept")))
+int sock_accept(int fd, int fdflags, int* result_ptr);
+
+
+static int
+wasi_transport_recv (void *buf, int len)
+{
+	int res;
+	int total = 0;
+	int fd = conn_fd;
+	int again = 0;
+
+	do {
+		res = read (fd, (char *) buf + total, len - total);
+		if (res > 0)
+			total += res;
+		again++;
+	} while ((res > 0 && total < len) || (res == -1 && again < retry_receive_message));
+	return total;
+}
+
+static gboolean
+wasi_transport_send (void *data, int len)
+{
+	int res;
+	do {
+		res = write (conn_fd, (const char*)data, len);
+	} while (res == -1);
+
+	if (res != len)
+		return FALSE;
+	else
+		return TRUE;
+	return TRUE;
```

Comment:
unreachable code. Why not `return res == len` ?

Reply:
done. :)